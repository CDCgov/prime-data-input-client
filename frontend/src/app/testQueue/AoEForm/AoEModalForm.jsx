import React from "react";
import { displayFullName } from "../../utils";
import CMSDialog from "../../commonComponents/CMSDialog";

import "@cmsgov/design-system/dist/css/index.css";
import { useState } from "react";
import {
  getSymptomList,
  getTestTypes,
  getPregnancyResponses,
} from "./constants";
import RadioGroup from "../../commonComponents/RadioGroup";
import Dropdown from "../../commonComponents/Dropdown";
import Anchor from "../../commonComponents/Anchor";
import Button from "../../commonComponents/Button";
import { COVID_RESULTS } from "../../constants";
import DateInput from "../../commonComponents/DateInput";

// building blocks that *probably* don't want to be exportable components
const SymptomInputs = ({
  symptomListConfig,
  noSymptoms,
  setNoSymptoms,
  currentSymptoms,
  setSymptoms,
  onsetDate,
  setOnsetDate,
}) => {
  const symptomChange = (evt) => {
    const choice = evt.currentTarget;
    setSymptoms({ ...currentSymptoms, [choice.value]: choice.checked });
  };
  const choiceList = symptomListConfig.map((conf) => {
    const { label, value } = conf;
    return { label, value, checked: currentSymptoms[value] };
  });

  // build the choice list we would have without the no-symptom toggle,
  // so we can extract its contents to plug into the wrapper.
  const wrappedSymptomChoiceList = noSymptoms ? null : (
    <React.Fragment>
      <hr />
      <RadioGroup
        type="checkbox"
        name="symptom_list"
        onChange={symptomChange}
        buttons={choiceList}
      />
      <DateInput
        label="Date of symptom onset"
        name="symptom_onset"
        value={onsetDate}
        onChange={setOnsetDate}
        max={new Date().toISOString().split("T")[0]}
      />
    </React.Fragment>
  );

  return (
    <React.Fragment>
      <RadioGroup
        type="checkbox"
        name="symptom_list"
        legend="Are you experiencing any of the following symptoms?"
        displayLegend
        onChange={(evt) => {
          console.log("new value:", evt.currentTarget.checked);
          setNoSymptoms(evt.currentTarget.checked);
        }}
        buttons={[{ value: "hasNoSymptoms", label: "No Symptoms" }]}
        selectedRadio={noSymptoms ? "hasNoSymptoms" : null}
      />
      {wrappedSymptomChoiceList}
    </React.Fragment>
  );
};

const PriorTestInputs = ({
  testTypeConfig,
  isFirstTest,
  setIsFirstTest,
  priorTestDate,
  setPriorTestDate,
  priorTestResult,
  setPriorTestResult,
  priorTestType,
  setPriorTestType,
}) => {
  // disable inputs other than Yes/No if it is "Yes" or unset
  const disableDetails = isFirstTest !== false;

  let radioValue = isFirstTest
    ? "yes"
    : isFirstTest === false
    ? "no"
    : undefined;

  return (
    <React.Fragment>
      <RadioGroup
        buttons={[
          { label: "Yes", value: "yes" },
          { label: "No", value: "no" },
        ]}
        selectedRadio={radioValue}
        onChange={(e) => {
          let value = e.target.value;
          let isFirstTest = value === "yes" ? true : false;
          setIsFirstTest(isFirstTest);
        }}
        legend="First covid test?"
        displayLegend
        name="prior_test_flag"
        horizontal
      />
      {isFirstTest ? null : (
        <>
          <DateInput
            label="Date of most recent prior test?"
            name="prior_test_date"
            value={priorTestDate}
            onChange={setPriorTestDate}
            disabled={disableDetails}
            max={new Date().toISOString().split("T")[0]}
            min="2020-02-01"
          />
          <Dropdown
            options={testTypeConfig}
            label="Type of prior test"
            name="prior_test_type"
            selectedValue={priorTestType}
            onChange={(e) => setPriorTestType(e.target.value)}
            disabled={disableDetails}
          />
          <Dropdown
            options={[
              { value: COVID_RESULTS.POSITIVE, label: "Positive" },
              { value: COVID_RESULTS.NEGATIVE, label: "Negative" },
              { value: COVID_RESULTS.INCONCLUSIVE, label: "Undetermined" },
            ]}
            label="Result of prior test"
            name="prior_test_result"
            selectedValue={priorTestResult}
            onChange={(e) => setPriorTestResult(e.target.value)}
            disabled={disableDetails}
          />
        </>
      )}
    </React.Fragment>
  );
};

const AoEModalForm = ({
  saveButtonText = "Save",
  onClose,
  patient,
  loadState = {},
  saveCallback = () => null,
}) => {
  const testConfig = getTestTypes();
  const symptomConfig = getSymptomList();

  // this seems like it will do a bunch of wasted work on re-renders and non-renders,
  // but it's all small-ball stuff for now
  const initialSymptoms = {};
  if (loadState.symptoms) {
    const loadedSymptoms = JSON.parse(loadState.symptoms);

    symptomConfig.forEach((opt) => {
      const val = opt.value;
      if (typeof loadedSymptoms[val] === "string") {
        initialSymptoms[val] = loadedSymptoms[val] === "true" || false;
      } else {
        initialSymptoms[val] = loadedSymptoms[val];
      }
    });
  } else {
    symptomConfig.forEach((opt) => {
      const val = opt.value;
      initialSymptoms[val] = false;
    });
  }

  const [noSymptoms, setNoSymptoms] = useState(loadState.noSymptoms || false);
  const [currentSymptoms, setSymptoms] = useState(initialSymptoms);
  const [onsetDate, setOnsetDate] = useState(loadState.symptomOnset);

  const [isFirstTest, setIsFirstTest] = useState(loadState.firstTest);
  const [priorTestDate, setPriorTestDate] = useState(loadState.priorTestDate);
  const [priorTestType, setPriorTestType] = useState(loadState.priorTestType);
  const [priorTestResult, setPriorTestResult] = useState(
    loadState.priorTestResult
  );

  const [pregnancyResponse, setPregnancyResponse] = useState(
    loadState.pregnancy
  );

  const saveAnswers = (evt) => {
    const saveSymptoms = { ...currentSymptoms };
    if (noSymptoms) {
      // set all symptoms explicitly to false
      symptomConfig.forEach(({ value }) => {
        saveSymptoms[value] = false;
      });
    }
    const priorTest = isFirstTest
      ? {
          firstTest: true,
          priorTestDate: null,
          priorTestType: null,
          priorTestResult: null,
        }
      : {
          firstTest: false,
          priorTestDate: priorTestDate,
          priorTestType: priorTestType,
          priorTestResult: priorTestResult,
        };

    saveCallback({
      noSymptoms,
      symptoms: JSON.stringify(saveSymptoms),
      symptomOnset: onsetDate,
      ...priorTest,
      pregnancy: pregnancyResponse,
    });
    onClose();
  };

  const actionButtons = (
    <div style={{ float: "right" }}>
      <Anchor text="Cancel" onClick={onClose} />
      <Button label={saveButtonText} onClick={saveAnswers} />
    </div>
  );
  const patientName = displayFullName(
    patient.firstName,
    patient.middleName,
    patient.lastName
  );
  return (
    <CMSDialog
      onExit={onClose}
      closeText="Cancel"
      heading={patientName}
      getApplicationNode={() => {
        document.getElementById("#root");
      }}
      actions={actionButtons}
      actionsInHeader={true}
    >
      <h2>Symptoms</h2>
      <SymptomInputs
        noSymptoms={noSymptoms}
        setNoSymptoms={setNoSymptoms}
        currentSymptoms={currentSymptoms}
        setSymptoms={setSymptoms}
        symptomListConfig={symptomConfig}
        setOnsetDate={setOnsetDate}
        onsetDate={onsetDate}
      />

      <h2>Past Tests</h2>
      <PriorTestInputs
        testTypeConfig={testConfig}
        priorTestDate={priorTestDate}
        setPriorTestDate={setPriorTestDate}
        isFirstTest={isFirstTest}
        setIsFirstTest={setIsFirstTest}
        priorTestType={priorTestType}
        setPriorTestType={setPriorTestType}
        priorTestResult={priorTestResult}
        setPriorTestResult={setPriorTestResult}
      />

      <h2>Pregnancy</h2>
      <RadioGroup
        legend="Currently pregnant?"
        displayLegend
        name="pregnancy"
        type="radio"
        onChange={(evt) => setPregnancyResponse(evt.currentTarget.value)}
        buttons={getPregnancyResponses()}
        selectedRadio={pregnancyResponse}
      />
    </CMSDialog>
  );
};

AoEModalForm.propTypes = {};

export default AoEModalForm;
