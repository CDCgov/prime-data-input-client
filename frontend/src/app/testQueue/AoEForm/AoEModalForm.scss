.sr-time-of-test-footer {
  margin-top: 2em;
}
.sr-time-of-test-buttons button {
  margin-left: 1em;
}
