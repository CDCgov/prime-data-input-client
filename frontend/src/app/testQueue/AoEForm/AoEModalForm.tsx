import React, { useRef, useState } from "react";
import QRCode from "react-qr-code";
import Modal from "react-modal";
import AoEForm from "./AoEForm";
import Button from "../../commonComponents/Button";
import RadioGroup from "../../commonComponents/RadioGroup";
import { displayFullName } from "../../utils";
import { globalSymptomDefinitions } from "../../../patientApp/timeOfTest/constants";
import { getUrl } from "../../utils/url";
import { gql, useQuery } from "@apollo/client";

interface LastTestData {
  patient: {
    lastTest: {
      dateTested: string;
      result: string;
    };
  };
}

export const LAST_TEST_QUERY = gql`
  query GetPatientsLastResult($patientId: String!) {
    patient(id: $patientId) {
      lastTest {
        dateTested
        result
      }
    }
  }
`;

interface Props {
  saveButtonText?: string;
  onClose: () => void;
  patient: any;
  loadState?: any;
  saveCallback: (a: any) => string | void;
  qrCodeValue?: string;
}

const AoEModalForm = (props: Props) => {
  const {
    saveButtonText = "Continue",
    onClose,
    patient,
    loadState = {},
    saveCallback,
    qrCodeValue,
  } = props;

  const [modalView, setModalView] = useState("");
  const [patientLink, setPatientLink] = useState("");
  const formRef = useRef<HTMLFormElement>(null);
  const modalViewValues = [
    { label: "Complete on smartphone", value: "smartphone" },
    { label: "Complete questionnaire verbally", value: "verbal" },
  ];

  const symptomsResponse: { [key: string]: boolean } = {};
  globalSymptomDefinitions.forEach(({ value }) => {
    symptomsResponse[value] = false;
  });

  const patientResponse = {
    firstTest: false,
    noSymptoms: false,
    pregnancy: undefined,
    priorTestDate: undefined,
    priorTestResult: null,
    priorTestType: undefined,
    symptomOnset: undefined,
    symptoms: JSON.stringify(symptomsResponse),
  };

  const continueModal = () => {
    // No need to save form if in "smartphone" mode
    if (modalView === "smartphone") {
      return onClose();
    }
    // Save default if form doesn't exist, otherwise submit form
    if (!formRef?.current) {
      saveCallback(patientResponse);
    } else {
      formRef.current.dispatchEvent(new Event("submit"));
    }
    onClose();
  };

  const chooseModalView = async (view: string) => {
    if (view === "smartphone") {
      // if we already have a truthy qrCodeValue, we do not need to save the test order to generate a PLID
      setPatientLink(
        qrCodeValue ||
          `${getUrl()}pxp?plid=${await saveCallback(patientResponse)}`
      );
    }
    setModalView(view);
  };

  const buttonGroup = (
    <div className="sr-time-of-test-buttons">
      <Button variant="unstyled" label="Cancel" onClick={onClose} />
      <Button
        className="margin-right-0"
        label={saveButtonText}
        type={"button"}
        onClick={() => continueModal()}
      />
    </div>
  );

  const { data, loading, error } = useQuery<LastTestData, {}>(LAST_TEST_QUERY, {
    fetchPolicy: "no-cache",
    variables: { patientId: patient.internalId },
  });
  if (loading) return null;
  if (error) throw error;
  const lastTest = data?.patient.lastTest;

  return (
    <Modal
      isOpen={true}
      style={{
        content: {
          maxHeight: "90vh",
          width: "40em",
          position: "initial",
        },
      }}
      overlayClassName="prime-modal-overlay display-flex flex-align-center flex-justify-center"
      contentLabel="Time of Test Questions"
    >
      <div className="display-flex flex-justify">
        <h1 className="font-heading-lg margin-top-05 margin-bottom-0">
          {displayFullName(
            patient.firstName,
            patient.middleName,
            patient.lastName
          )}
        </h1>
        {buttonGroup}
      </div>
      <div className="border-top border-base-lighter margin-x-neg-205 margin-top-205"></div>
      {process.env.REACT_APP_PATIENT_EXPERIENCE_ENABLED === "true" ? (
        <>
          <h2 className="font-heading-lg margin-top-205 margin-bottom-0">
            Test questionnaire
          </h2>
          <RadioGroup
            legend="How would you like to complete the questionnaire?"
            name="qr-code"
            type="radio"
            onChange={(evt) => chooseModalView(evt.currentTarget.value)}
            buttons={modalViewValues}
            selectedRadio={modalView}
            className="margin-top-205"
          />
          {modalView === "smartphone" && (
            <>
              <section className="display-flex flex-justify-center margin-top-4 padding-top-5 border-top border-base-lighter">
                <div className="text-center">
                  <p className="font-body-lg margin-y-0">
                    Point your camera at the QR code <br />
                    to access the questionnaire
                  </p>
                  <div
                    className="margin-top-205"
                    id="patient-link-qr-code"
                    data-patient-link={patientLink}
                  >
                    <QRCode value={patientLink} size={190} />
                  </div>
                </div>
              </section>
              <div className="border-top border-base-lighter margin-x-neg-205 margin-top-5 padding-top-205 text-right">
                <Button
                  className="margin-right-205"
                  label={saveButtonText}
                  type={"button"}
                  onClick={() => continueModal()}
                />
              </div>
            </>
          )}
          {modalView === "verbal" && (
            <AoEForm
              saveButtonText="Continue"
              onClose={onClose}
              patient={patient}
              lastTest={lastTest}
              loadState={loadState as any}
              saveCallback={saveCallback}
              isModal={true}
              noValidation={true}
              formRef={formRef}
            />
          )}
        </>
      ) : (
        <AoEForm
          saveButtonText="Continue"
          onClose={onClose}
          patient={patient}
          loadState={loadState as any}
          saveCallback={saveCallback}
          lastTest={lastTest}
          isModal={true}
          noValidation={true}
          formRef={formRef}
        />
      )}
    </Modal>
  );
};

AoEModalForm.propTypes = {};

export default AoEModalForm;
