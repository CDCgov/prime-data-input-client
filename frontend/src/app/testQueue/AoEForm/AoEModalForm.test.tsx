import renderer, { act } from "react-test-renderer";
import MockDate from "mockdate";

import AoEModalForm from "./AoEModalForm";
import ReactDOM from "react-dom";

jest.mock("./AoEForm", () => () => <></>);
jest.mock("react-modal", () => (props: any) => <>{props.children}</>);

describe("AoEModalForm", () => {
  let component: renderer.ReactTestRenderer;

  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((element, node) => {
      return element;
    }) as any;
  });

  beforeEach(() => {
    MockDate.set("2021-02-06");
    component = renderer.create(
      <div id="test-modal-container">
        <AoEModalForm
          saveButtonText="save"
          onClose={jest.fn()}
          patient={{
            internalId: "123",
            gender: "male",
          }}
          loadState={{
            noSymptoms: false,
            symptoms: '{"426000000":"true","49727002":false}',
            symptomOnset: "",
            priorTestDate: "",
            priorTestResult: "",
            priorTestType: "",
            firstTest: false,
            pregnancy: "",
          }}
          saveCallback={jest.fn()}
        />
      </div>
    );
  });

  describe("on data loaded", () => {
    beforeEach(async () => {
      // load data
      await act(async () => {
        await new Promise((resolve) => setTimeout(resolve, 0));
      });
    });
    it("renders", async () => {
      expect(component.toJSON()).toMatchSnapshot();
    });
  });
});
