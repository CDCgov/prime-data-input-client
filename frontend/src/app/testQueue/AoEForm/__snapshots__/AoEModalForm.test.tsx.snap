// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AoEModalForm on data loaded has a distinct view for 'text' 1`] = `
<div
  id="test-modal-container"
>
  <div
    className="display-flex flex-justify"
  >
    <h1
      className="font-heading-lg margin-top-05 margin-bottom-0"
    >
      Jobs, Steve
    </h1>
    <div
      className="sr-time-of-test-buttons"
    >
      <button
        className="usa-button usa-button--unstyled"
        onClick={[MockFunction]}
        type="button"
      >
        Cancel
      </button>
      <button
        className="usa-button margin-right-0"
        onClick={[Function]}
        type="button"
      >
        save
      </button>
    </div>
  </div>
  <div
    className="border-top border-base-lighter margin-x-neg-205 margin-top-205"
  />
  <h2
    className="font-heading-lg margin-top-205 margin-bottom-0"
  >
    Test questionnaire
  </h2>
  <fieldset
    className="usa-fieldset prime-radios margin-top-205"
  >
    <legend
      className="usa-legend"
    >
      How would you like to complete the questionnaire?
      <span
        className="usa-hint"
      >
         (optional)
      </span>
    </legend>
    <div
      className="usa-form-group"
    >
      <div
        className="usa-radio"
      >
        <input
          checked={true}
          className="usa-radio__input"
          data-required="false"
          disabled={false}
          id="radio-test-1_text_0"
          name="qr-code"
          onChange={[Function]}
          type="radio"
          value="text"
        />
        <label
          className="usa-radio__label"
          htmlFor="radio-test-1_text_0"
        >
          Text message
          <span
            className="radio__label-description--checked usa-radio__label-description text-base"
          />
        </label>
      </div>
      <div
        className="usa-radio"
      >
        <input
          checked={false}
          className="usa-radio__input"
          data-required="false"
          disabled={false}
          id="radio-test-1_verbal_1"
          name="qr-code"
          onChange={[Function]}
          type="radio"
          value="verbal"
        />
        <label
          className="usa-radio__label"
          htmlFor="radio-test-1_verbal_1"
        >
          Complete questionnaire verbally
        </label>
      </div>
    </div>
  </fieldset>
  <div
    className="border-top border-base-lighter margin-x-neg-205 margin-top-5 padding-top-205 text-right"
  >
    <button
      className="usa-button margin-right-205"
      onClick={[Function]}
      type="button"
    >
      Text link to complete on smartphone
    </button>
  </div>
</div>
`;

exports[`AoEModalForm on data loaded has a distinct view for 'verbal' 1`] = `
<div
  id="test-modal-container"
>
  <div
    className="display-flex flex-justify"
  >
    <h1
      className="font-heading-lg margin-top-05 margin-bottom-0"
    >
      Jobs, Steve
    </h1>
    <div
      className="sr-time-of-test-buttons"
    >
      <button
        className="usa-button usa-button--unstyled"
        onClick={[MockFunction]}
        type="button"
      >
        Cancel
      </button>
      <button
        className="usa-button margin-right-0"
        onClick={[Function]}
        type="button"
      >
        save
      </button>
    </div>
  </div>
  <div
    className="border-top border-base-lighter margin-x-neg-205 margin-top-205"
  />
  <h2
    className="font-heading-lg margin-top-205 margin-bottom-0"
  >
    Test questionnaire
  </h2>
  <fieldset
    className="usa-fieldset prime-radios margin-top-205"
  >
    <legend
      className="usa-legend"
    >
      How would you like to complete the questionnaire?
      <span
        className="usa-hint"
      >
         (optional)
      </span>
    </legend>
    <div
      className="usa-form-group"
    >
      <div
        className="usa-radio"
      >
        <input
          checked={false}
          className="usa-radio__input"
          data-required="false"
          disabled={false}
          id="radio-test-1_text_0"
          name="qr-code"
          onChange={[Function]}
          type="radio"
          value="text"
        />
        <label
          className="usa-radio__label"
          htmlFor="radio-test-1_text_0"
        >
          Text message
          <span
            className="radio__label-description--checked usa-radio__label-description text-base"
          />
        </label>
      </div>
      <div
        className="usa-radio"
      >
        <input
          checked={true}
          className="usa-radio__input"
          data-required="false"
          disabled={false}
          id="radio-test-1_verbal_1"
          name="qr-code"
          onChange={[Function]}
          type="radio"
          value="verbal"
        />
        <label
          className="usa-radio__label"
          htmlFor="radio-test-1_verbal_1"
        >
          Complete questionnaire verbally
        </label>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`AoEModalForm on data loaded renders 1`] = `
<div
  id="test-modal-container"
>
  <div
    className="display-flex flex-justify"
  >
    <h1
      className="font-heading-lg margin-top-05 margin-bottom-0"
    >
      Jobs, Steve
    </h1>
    <div
      className="sr-time-of-test-buttons"
    >
      <button
        className="usa-button usa-button--unstyled"
        onClick={[MockFunction]}
        type="button"
      >
        Cancel
      </button>
      <button
        className="usa-button margin-right-0"
        onClick={[Function]}
        type="button"
      >
        save
      </button>
    </div>
  </div>
  <div
    className="border-top border-base-lighter margin-x-neg-205 margin-top-205"
  />
  <h2
    className="font-heading-lg margin-top-205 margin-bottom-0"
  >
    Test questionnaire
  </h2>
  <fieldset
    className="usa-fieldset prime-radios margin-top-205"
  >
    <legend
      className="usa-legend"
    >
      How would you like to complete the questionnaire?
      <span
        className="usa-hint"
      >
         (optional)
      </span>
    </legend>
    <div
      className="usa-form-group"
    >
      <div
        className="usa-radio"
      >
        <input
          checked={false}
          className="usa-radio__input"
          data-required="false"
          disabled={false}
          id="radio-test-1_text_0"
          name="qr-code"
          onChange={[Function]}
          type="radio"
          value="text"
        />
        <label
          className="usa-radio__label"
          htmlFor="radio-test-1_text_0"
        >
          Text message
          <span
            className="radio__label-description--checked usa-radio__label-description text-base"
          />
        </label>
      </div>
      <div
        className="usa-radio"
      >
        <input
          checked={false}
          className="usa-radio__input"
          data-required="false"
          disabled={false}
          id="radio-test-1_verbal_1"
          name="qr-code"
          onChange={[Function]}
          type="radio"
          value="verbal"
        />
        <label
          className="usa-radio__label"
          htmlFor="radio-test-1_verbal_1"
        >
          Complete questionnaire verbally
        </label>
      </div>
    </div>
  </fieldset>
</div>
`;
